### ISSUE TITLE
Making movebase work

### ISSUE THREAD

**Issue by marc-hanheide** [strands-project/strands_movebase#62]
team members @cburbridge @bfalacerda @cdondrup @Pandoro @Jailander @denisehe @TobKoer @PDuckworth 
## Identified issues
1. global planner is updating too slowly (tried 10Hz)
2. the local planner's parameterisation is off and uses many "guessed" values
3. we need a test definition (to have some "genetic" optimisation for the parameters)
4. observed problems
   1. turning towards obstacles
   2. stopping too late (maybe acceleration are just wrong)
   3. zig-zag paths
   4. making physical contact
   5. the robot always tries to go to the exact waypoint (should already be relaxed with the newer version)
   6. local cost map probably too small
   7. global cost map inflation could be wrong: We don't want the robot to be always in the middle of a corridor, so we don't want any gradual inflation. 
   8. if the robot cannot reach a waypoint (closed door), it stays in front of the door blocking it (might be solved better on topological / execution level)
   9. robot is quite "pushy" navigating through crowds (this should already be working with human-aware nav)
## Test cases
- general setup: L-shape, with 3 topo map with 3 waypoints (one intermediate)
- running via topo nav
- collect rosbags
- the obstacles are always "around the corner", 1m away from intersection
### static tests
1. place robot close to obstacle (10cm away, 45, 90, 135), goals right behind, and at 45 degree angle
2. traverse narrow corridor (1m, corridor in global map)
3. traverse corridor with row of chairs (2m wide, leaving a 1m gap)
   a. all chairs on one side
   b. chairs on both sides
4. robot trapped 
5. (optional: traverse a door)
6. blocking the exact intermediate waypoint (putting a chair there)
7. blocking the target waypoint (putting a wheelchair there), stopping at appropriate distance
8. chairs blocking the whole route
9. human standing next to wall in 2m corridor, with crutches sticking out, creating a gap of about 1m
### dynamic tests
1. emergency stop: "jump" in front of robot (50 cm distance)
2. pass-crossing: (person walks / wheelchair drives) in front to the robot coming from the side: robot should stop
3. pass-by: (person walks / wheelchair drives) in front to the robot coming from the side: robot should stop
4. pass-crossing and pass-by with a group of people (group blocks the way for the robot), with a 10 second break of the group standing in front of robot
## Thinks to do
- [x] collect all nav params used across sites and comment your observation / characteristic (please comment on this issue)
- [ ] look at cost maps in the mon nav from deployment data and identify the situations where it failed
- [ ] implement above test cases and iterate parameters on all sites

To prevent regressions:
- [ ] implement navigation unit tests, based on ROS bag
- [x] implement navigation unit tests, based on Morse
## Further ideas:
- make it more clearly when the robot is about to move
- put warning lights on to indicate when it is about to move (complementing speech)
- check for RAM/CPU usage
- watch dogs need to be always running

**Comment by marc-hanheide** [comment]
OK, so this is the brief summary of what we planned to do. No it's time to get going on this. 
Can ask representatives of the sites (maybe @bfalacerda @Pandoro @Jailander @PDuckworth and @nilsbore) to check the parameters they are using, post them (i.e. link them) from here and write one paragraph summarising your observations.

Then, we should create the setup for the test cases, e.g. have a shared topological map which we only slightly adapt to our own environments, but have the names consistent. I suggest, UOL (@jailander, myself, @cdondrup) create a first prototype of the test setup.

Then we should have a hangout to discuss the next steps. Here's a [doodle](https://doodle.com/poll/56f7c5ydtqny8p78) I ask you all to fill to arrange this.

Many thanks

**Comment by marc-hanheide** [comment]
forgot @hawesie... sorry

**Comment by bfalacerda** [comment]
I forgot we can't check what params we were using on Bob during the deployment because his hard drive died. I'm pretty sure we were using the current strands_movebase ones though.

However, @kunzel did some changes to the dwa params  to make it follow the global path more closely. It was only added to [scitos_2d_nav](https://github.com/strands-project/scitos_2d_navigation/blob/hydro-devel/scitos_move_base_params/dwa_planner_ros.yaml), because it makes navigation in simulation work much better. The differences are 
- sim_time: 0.8 (was 2.1) This one was done because the dwa scoring only looks at the end of the sampled trajectories for 2 of the components of the scoring function. This made it choose paths that were two far away from what the global planner wanted:

```
cost =
  path_distance_bias * (distance to path from the endpoint of the trajectory in meters)
  + goal_distance_bias * (distance to local goal from the endpoint of the trajectory in meters)
  + occdist_scale * (maximum obstacle cost along the trajectory in obstacle cost (0-254))
```
- path_distance_bias: 1.0 (was 32.0, the default),  goal_distance_bias: 0.5 (was 24.0, the default). This has to do with the same issue as above I think

Maybe @cdondrup has some insight on these params, since he has been looking at dwa?

**Comment by cdondrup** [comment]
Actually setting `path_distance_bias` to 1 and `goal_distance_bias` to 0.5 should have had the opposite effect and allow the robot to diverge from the global path a bit more. However, if the DWA only uses the params described above (I remember there was goal align and path align somewhere as well but that might have been in a different implementations of it) then this shouldn't have changed much. The ratio is key here so with 32 and 24 it should have been the same as 1.0 and 0.75... so by changing it to 1.0 and 0.5 we only just gave less weight to the `goal_distance_bias`.

The `occdist_scale` is the one responsible for obstacle avoidance and the default here should be 0.01 which means it only takes lethal obstacles into account and doesn't have effect on the other two params.

**Comment by bfalacerda** [comment]
yes, from what i understand, the short sim time is what makes a big difference here: since both those values only look at the endpoint of the simulated trajectories, if we increase their size we increase the number of trajectories that don't go along the global plan, but have an endpoint close to it

**Comment by cdondrup** [comment]
Isn't that what you want for obstacle avoidance?

**Comment by bfalacerda** [comment]
depends. if we increase the rate of the global planner it might be better to keep the sim time smaller? with this version we had the robot rotating a lot more in place, at least in simulation.

**Comment by Pandoro** [comment]
We are indeed using some custom parameters. We had our old hiwi run around and optimize the parameters in such a way that the behaviour would be 'optimal' and we have the following in our "site parameter"-file:

``` yaml
local_costmap:
  width: 4
  height: 4
  resolution: 0.02

  # Improves navigation in tight corridors tremendously.
  inflation_layer:
    inflation_radius: 0.5

DWAPlannerROS:
  max_vel_y: 0.0
  min_vel_y: 0.0
```

I can't really say what the idea was behind these changes, just that it felt a bit more stable after he played around with the parameters. In general navigation seems okayish, apart from of course the driving towards obstacles and the slightly awkward behaviour in our kitchen that I described during the GA. According to our launch file we are supposedly running the human aware navigation, but he doesn't stare at people so maybe something went going wrong there. I'll check more accurately on our next run with Karl.

**Comment by marc-hanheide** [comment]
Could you please submit a link to the "whole" set of files, as this already assumes there would be a standard you deviate from ;-)

**Comment by Pandoro** [comment]
You can check out the [strands karl repo](https://github.com/lucasb-eyer/strands_karl). This should have everything we use, including the above file. We took most of the stuff from what we had before in the marathon system and updated it with everything new and relevant from the review system.

**Comment by marc-hanheide** [comment]
@Pandoro @Jailander @cdondrup Can I ask you to fill the doodle poll at https://doodle.com/poll/56f7c5ydtqny8p78 as well, please?

**Comment by marc-hanheide** [comment]
For the record, I link the current versions of params here:

## BHAM
- normally based on [strands_movebase](https://github.com/strands-project/strands_movebase/tree/b30a6e3fc85745a3161c8d61feaa720262b450d2/strands_movebase/strands_movebase_params)
- with some changes dicussed in https://github.com/strands-project/strands_movebase/issues/62#issuecomment-148691500

## RWTH
- seems pretty much standard, but this is the [used version](https://github.com/lucasb-eyer/strands_karl/tree/ad813f614fd6d065bd3bcb7b1322f547953ea555/launch)

## UOL

missing

## KTH
- standard parameters

**Comment by marc-hanheide** [comment]
@bfalacerda do you remember why we move to our own NavFn? https://github.com/strands-project/strands_movebase/commit/561836295f0418981240e0b1eb5db40f90d7c0ba

**Comment by bfalacerda** [comment]
yes, so we could reconfigure the default_tolerance param a runtime. This  makes the robot not go to the exact pose of the intermediary waypoints.

**Comment by marc-hanheide** [comment]
right, I remember...

**Comment by nilsbore** [comment]
A git diff to the branch we are using reveals no changes to the parameters.

**Comment by Jailander** [comment]
@marc-hanheide in Linda we use the standard parameters namely the ones at https://github.com/strands-project/strands_movebase/tree/indigo-devel/strands_movebase/strands_movebase_params

**Comment by marc-hanheide** [comment]
simulation env in https://github.com/strands-project/strands_morse/pull/138

**Comment by nilsbore** [comment]
http://answers.ros.org/question/201535/move_basedwa-parameters-for-a-large-robot-with-rotation-center-in-the-front/

> Then I thought well, I am going to read the source code, which helped a lot. and thanks for good coding + comments. In dwa_local_planner page, the forward_point_distance is defined as an additional scoring point in front of the robot. I frankly didn't understand what it meant. However, in the source code, I found out that this value is being used as a cost value for robot alignment to the path. By setting it to zero, the trajectory will not be penalized if the robot is not aligned on the path. This is a very important factor for differential drive robots whose centre of rotation is not the in the middle of the robot. Apparently, a lot of valid trajectories were being dropped because of this. As soon as I set this to zero, the robot was able to perform much better. However, improvements could be done so I continued.

This makes me want to investigate how the robot behaves if we set `forward_point_distance: 0.0`. The point afterwards about visualizing the evaluated trajectories also sounds like a good point.

**Comment by marc-hanheide** [comment]
very useful @nilsbore !

Quick summary of today's hangout:
1. @Jailander to define the first test case as part of strands_navigation unit test as a blueprint
2. @bfalacerda to talk to @cburbridge to fork https://github.com/ros-planning/navigation into STRANDS including Chris' visualisation of policies (and release it from there)
3. @cdondrup to work on the testing framework, including recording of trajectories. It has been discussed that we split the test into a test preparation service which will differ in simulation and real, and an actual unit test that call the preparation service (e.g. to "beam" the robot to its starting position), and then run the actual test.
4. Everybody to replicate first test case "live" once working
5. Everybody to add additional unit tests for the other cases identified above (incrementally, always in line with live robot testing)
6. @marc-hanheide to facilitate simulation test on Jenkins: https://github.com/strands-project/strands_ci/issues/15
7. @creuther to check again for relation between force and DWA parameters

**Comment by cburbridge** [comment]
I have forked navigation into strands-project and added my debug out changes to the "strands-testing" branch. These changes are not permanent as they make the code less efficient, so this branch should only be used for debugging etc. Therefore I don't think a "release" would be a good idea.

To use it, download and build the repo in you workspace then restart navigation, and run the (ugly) python script dwa_planner/scripts/log.py. This will open a window with six images, each showing the score for the critics used in the code. The axis for each graph are: y=angular velocity, x=linear velocity, centre is zero and the values are scaled by 10. so (x=10,y=30) represents (lv=0, wv=0), (x=15,y=20) represents (lv=0.5, wv=-0.1).  The graphs are updated in real time as the robot drives, but some timesteps will be missed due to the speed of the redraw. If this is a problem then you can log the topic debug topic and replay it slower...

**Comment by marc-hanheide** [comment]
@cburbridge agreed, not to be released. I just thought we'll eventually fix code in navigation and then we'd have our own. But for now, let's leave it as a debug tool.

**Comment by marc-hanheide** [comment]
People will (probably) like to hear that Jenkins can now run simulation-enabled test (more precisely, it can run test in a virtual OpenGL/X environment, hence allowing us to write unit test that use the simulator and/or other GUIs. https://github.com/strands-project/strands_navigation/pull/270 is a first example (though a stupid one as it only starts the simulator but doesn't do anything with it, but it's a test). One outcome of any tests we now run on jenkins (devel or PR) is now a highly compressed (1 fps) video (see https://lcas.lincoln.ac.uk/jenkins/view/Pull%20Requests/job/pr-indigo-strands_navigation/98/artifact/Xvnc.mp4 for the corresponding example) showing the virtual desktop (jump to minute 6 in said video to see the simulator appearing). This video should help to identify causes of failure in unit tests.
As this is facilitated by a virtualised compute server GPU, you cannot expect the performance of a real GPU here, but it's enough to run morse in wireframe (and maybe also small/simple environments).

Having written all this, I should probably write the same to the mailing list...

Anyway, so the infrastructure is there to actually write more elaborate tests... now somebody needs to do it ;-)

**Comment by creuther** [comment]
I talked to my people and this is the (probably not as satisfying as hoped for) gist of it:

**1**. The SCITOSDriver specifies a whole lot of maximum/minimum velocities and maximum positive/negative accelerations. These values are passed down to the MCU (Main Control Unit) and will be used as soft limits in the motor controller. Most of the time these constraints are met, but apparently it cannot be guaranteed (only in extreme situations).

This means that the "drive parameters" of all STRANDS robots are currently specified [here](https://github.com/strands-project/scitos_drivers/blob/indigo-devel/scitos_mira/resources/SCITOS-G5-model.xml).
More specifically, these parameters govern the way soft limits:

```
        <MaxForwardVelocity>1.0</MaxForwardVelocity>
        <MaxBackwardVelocity>-0.4</MaxBackwardVelocity>
        <MaxRotVelocity>120</MaxRotVelocity>
        <TransEpsilonVelocity>0.02</TransEpsilonVelocity>
        <RotEpsilonVelocity>1</RotEpsilonVelocity>
        <MaxTransAcceleration>0.8</MaxTransAcceleration>
        <MaxTransDeceleration>0.9</MaxTransDeceleration>
        <MaxRotAcceleration>180</MaxRotAcceleration>
        <MaxRotDeceleration>260</MaxRotDeceleration>
        <MaxEmergencyDeceleration>1.6</MaxEmergencyDeceleration>
```

Rotations given in degrees / {second,second^2} and translations given in meters / {second, second^2}.

The changes @bfalacerda made with respect to the DWA parameters should reflect those parameters. The parameters can also be tweaked to achieve softer acceleration, more abrupt braking, etc.

**2**. (paraphrasing Stefan, as I don't know a lot about electronics): The force basically imposes a threshold for the motor current, which is encoded as a 8 bit PWM (hence the values 0-255 cover the complete analogue spectrum). From a motor controlling perspective it would be ideal to always drive at full force, but due to wheel slippage and things like relying on the robot to get stopped by the charging station's threshold, we compromise and use a value of typically around 80-120. Therefore lowering the force too much will increase the stopping distance. Also, with a force of around 100 the acceleration/deceleration limits set (see above) aren't necessarily reached, so increasing the force will enable the robot to get closer to the set limits if they are "quite high".

There is no real formula or table I can give you regarding the force and the accelerations. But setting lower values for accelerations should mitigate the influence of changing the force on the acceleration/deceleration behaviour.

**Comment by marc-hanheide** [comment]
given the values are defined in the Mira config file would it make sense to expose them as rosparams and have DWA configured from those params? obviously we would expose them as rad/s(^2) where needed?

**Comment by creuther** [comment]
Certainly possible, they can be queried in MIRA and therefore exposed as rosparams. Though note that they are not dynamically reconfigurable, they are written to the MCU at startup and then can't be changed during runtime.
But still, it would probably beat having to maintain two sets of files (robot config file and dwa params file) and all the potential trouble that comes with that. I can get together with @cburbridge to help expose the properties.

**Comment by cburbridge** [comment]
I have done some test in the `scitos_mira` code, and I think there is a ~400ms delay between commanding a velocity and the odometry reporting that the robot is moving at that speed. This is  just the delay on the  MIRA side, it does not include any delay between the DWA selecting a velocity and the MIRA code sending that velocity to the firmware. @creuther would you expect this kind of timing?

**Comment by nilsbore** [comment]
I made some improvements to how the clouds are processed by movebase, which should gains us some more CPU on the main computer: https://github.com/strands-project/strands_movebase/pull/64 .

**Comment by marc-hanheide** [comment]
In response to https://github.com/strands-project/strands_movebase/issues/62#issuecomment-152651469: @creuther can you comment? A delay of almost half a second can indeed explain a lot of stuttering I guess. Is there a way to reduce this or is there a way to make DWA aware of such a delay (asking very silly questions)...

**Comment by creuther** [comment]
This seems very peculiar to me. I wanted to do a test on Betty but no one's here yet so I'm hesitant to stop the running ROS stuff. 

From our side, we expect some delay in setting and passing the command to the motors ( as always, giving an exact number is not possible, unfortunately... :( ) but **certainly** not 400ms. I'll have a look and report back as soon as someone turns up at the office ;)

**Comment by hawesie** [comment]
Tsk, lazy Brummies.

**Comment by cburbridge** [comment]
I'm away this week, but will look at it again on Monday.

**Comment by creuther** [comment]
Small update: Commanding a velocity from MIRA directly doesn't result in any significant delay, while the delay is definitely there if you command the velocity from ROS. I had a look through the scitos_mira code but can't really see any obvious points of concern.

The only thing that I am not too sure about is the following: [Here](https://github.com/strands-project/scitos_drivers/blob/indigo-devel/scitos_mira/src/ScitosG5.cpp#L36) we define that ScitosG5, i.e. pretty much the entire ROS-MIRA-bridge, should spin at 5Hz. Does the [following](https://github.com/strands-project/scitos_drivers/blob/indigo-devel/scitos_mira/src/ScitosG5.cpp#L41) sleep mean that the entire ROS-MIRA-bridge will be blocked for the remaining (200 - x) ms? I tried to set it to spin at 25Hz as well as 1Hz on Betty but didn't really make a difference. Then again @cburbridge 's debug timestamps show that the odometry is definitely received with 20Hz, so it really doesn't make a difference / block it as far as I can tell. Strange.

**Comment by cburbridge** [comment]
The 5Hz spin loop is a separate thread to the odometry message receiving,
so having this at 5Hz should be ok I think. However, there used to be
horrible bug in this respect a while back:
https://github.com/strands-project/scitos_drivers/pull/53 so this needs
more checking. Since the only component that registers functions to be
called in that loop is the head, it would be interesting to test without
that module. If you have time could you try running the scitos launch
again, but remove Head from
https://github.com/strands-project/scitos_drivers/blob/indigo-devel/scitos_mira/launch/scitos_mira.launch#L4.
If not no problem, I'll test it out on Monday.

**Comment by cdondrup** [comment]
Since the tests are almost finished, I have a few questions:

**Recovery Behaviours**: Which recovery behaviours, if any, do we want during the test runs? Especially things like backtrack make the whole test if DWA can turn when it faces a wall kind of pointless as the robot will backtrack and then turn. Since we want to tune DWA parameters, I think we have to be careful which recoveries to run. This brings me to another question: Why are there two different locations where the recovery behaviours are defined @bfalacerda ? I.e. in [the config file](https://github.com/strands-project/strands_recovery_behaviours/blob/hydro-devel/strands_recovery_behaviours/config/monitored_nav_config.yaml) and a [python file](https://github.com/strands-project/strands_recovery_behaviours/blob/hydro-devel/strands_monitored_nav_states/src/strands_monitored_nav_states/recover_nav.py).

**Results**: What feedback would like to get in addition to that it failed or succeeded. Currently, the service starting the test run will report if navigation to the goal was successful, if not, if the robot was able do die gracefully by at least being able to return to the starting location, the travelled distance and time. Once I took care of the human, I will also add the minimum distance kept to the human.

**Timeouts**: Currently, the navigation times out after 40 seconds and then tries to go back to the starting location for another 40 seconds. Meaning that if the goal is not reached in that time it will take the same time to try and go back. In total that means each test has a maximum of 80 seconds to at least die gracefully. Do we want a timeout at all or do we want to rely on navigation failing on its own? This might mean that the tests will run for quite a while or we have to reconfigure the sleep and retry argument (if we want that recovery behaviour at all) to make it fail quicker.

Maybe @marc-hanheide and/or @hawesie have an opinion on those.

**Comment by cdondrup** [comment]
Just briefly talked to @marc-hanheide and this is what we thought would be best:

**Recovery Behaviours**: None. This is not what we want to test here. Follow-up question for @bfalacerda by not giving the mon nav a config file it does not run the stuff in the config file obviously. For the ones in the python file I would have to disable them via the parameter?! Or can I disable them all together from the start. Maybe not even running monitored navigation for now?

**Timeouts**: No time outs. Navigation should fail on its own.

Any other thoughts on that?

**Comment by Jailander** [comment]
The config file tells which recovery behaviours python file to load, you need to run monitored navigation, but if you don't give any config file it will simply not load any (which is what we want in this case).

**Comment by bfalacerda** [comment]
I agree with @marc-hanheide, and @Jailander already answered the question about the config file :) You should also see a warning being printed telling you that there are no recovery behaviours added.

Regarding the config vs python, config defines the [RecoverStateMachine](https://github.com/strands-project/strands_navigation/blob/indigo-devel/monitored_navigation/src/monitored_navigation/recover_state_machine.py) to load. The SM itself is defined as a python class, since it's using smach. We need the python code to define how the SM moves between states.

**Comment by cdondrup** [comment]
We discovered a different problem regarding the timeout that I guess @bfalacerda solved somewhere in monitored nav that I can't find. Navigation never fails. Somewhere there has to be a timeout. Can you point me to that @bfalacerda ?

**Comment by bfalacerda** [comment]
I don't get the question. What do you mean navigation never fails?

There are no timeouts in monitored nav, it waits for move_base (or whatever action it calls) to give a result

**Comment by Jailander** [comment]
I think the question, is how to set a timeout in monitored navigation because move_base doesn't fail by itself

**Comment by bfalacerda** [comment]
it does, dwa eventually gives up. you can also  set the `planner_patience` and `controller_patience` params:
http://wiki.ros.org/move_base?distro=jade#Parameters

**Comment by cdondrup** [comment]
The problem is that if the robot starts in front of a wall (10cm distance) it just prints:

```
[ WARN] [1446723231.414625106]: DWA planner failed to produce path.
[ INFO] [1446723231.513456115]: Got new plan
```

until the end of times. The problem is that the robot is not inside an obstacle so navigation does not fail but is also not able to find a path. I let it run for a few minutes but it didn't give up. I'll have a look the link. Thanks!

**Comment by cdondrup** [comment]
OK, I don't think this will help much:

```
~planner_patience (double, default: 5.0)
How long the planner will wait in seconds in an attempt to find a valid plan before space-clearing operations are performed.
~controller_patience (double, default: 15.0)
How long the controller will wait in seconds without receiving a valid control before space-clearing operations are performed.
```

It does perform the clearing but then simply keeps on trying to no avail.

**Comment by bfalacerda** [comment]
are you sure it's not failing and being called over and over again? Are you using topo nav policy execution? I think it tries several times. monitored nav will always report back to topo after a failure so it can decide what to do

**Comment by bfalacerda** [comment]
the behaviour that move_base usually has after dwa failing for `controller_patience` seconds is clearing the costmaps and trying again. If valid controls are still not found, it should output aborted. I've never seen it getting stuck with dwa failing forever

**Comment by cdondrup** [comment]
Just to see if it does anything I ran it for 5 minutes (8 minutes now after finishing to write this) and the only thing it outputs is that DWA is unable to produce a path. This is the same when the robot, e.g. is surrounded by people, technically, there is nothing wrong because the robot is in free space and not inside an obstacle but the DWA can still not find a path. Hence, navigation will never fail and mon nav will never know.

It did clear the costmaps but then keeps on running till forever. I'm using the default move_base params:

``` yaml
cdondrup@calculon:~/ros_ws
$ rosparam get /move_base/
DWAPlannerROS: {acc_lim_th: 2.0, acc_lim_theta: 3.2, acc_lim_x: 1.0, acc_lim_y: 0.0,
  acc_limit_trans: 0.1, angular_sim_granularity: 0.1, forward_point_distance: 0.325,
  goal_distance_bias: 24.0, holonomic_robot: false, latch_xy_goal_tolerance: true,
  max_rot_vel: 1.0, max_scaling_factor: 0.2, max_trans_vel: 0.55, max_vel_x: 0.55,
  max_vel_y: 0.0, min_rot_vel: 0.4, min_trans_vel: 0.1, min_vel_x: 0.0, min_vel_y: 0.0,
  occdist_scale: 0.01, oscillation_reset_angle: 0.2, oscillation_reset_dist: 0.05,
  path_distance_bias: 32.0, prune_plan: true, restore_defaults: false, rot_stopped_vel: 0.1,
  scaling_speed: 0.25, sim_granularity: 0.025, sim_time: 2.1, stop_time_buffer: 0.2,
  trans_stopped_vel: 0.1, use_dwa: true, vth_samples: 20, vx_samples: 3, vy_samples: 10,
  xy_goal_tolerance: 0.3, yaw_goal_tolerance: 0.1}
NavfnROS: {allow_unknown: false}
aggressive_reset:
  layer_names: [obstacle_layer]
  reset_distance: 0.0
base_global_planner: navfn/NavfnROS
base_local_planner: dwa_local_planner/DWAPlannerROS
clearing_rotation_allowed: true
conservative_reset:
  layer_names: [obstacle_layer]
  reset_distance: 2.0
conservative_reset_dist: 3.0
controller_frequency: 10.0
controller_patience: 15.0
global_costmap:
  footprint: '[[-0.3,-0.32],[-0.52,-0.11],[-0.52,0.11],[-0.3,0.32],[-0.05,0.31],[0.05,0.25],[0.15,0.13],[0.2,0],[0.15,-0.13],[0.05,-0.25],[-0.05,-0.31]]'
  footprint_padding: 0.0001
  global_frame: /map
  height: 10
  inflation_layer: {cost_scaling_factor: 5.0, enabled: true, inflation_radius: 0.8}
  map_layer: {enabled: true, map_topic: /map}
  no_go_layer: {enabled: true, map_topic: /no_go_map}
  obstacle_layer:
    combination_method: 1
    enabled: true
    laser: {clearing: true, data_type: LaserScan, marking: true, sensor_frame: base_laser_link,
      topic: scan}
    mark_threshold: 0
    max_obstacle_height: 2.0
    observation_sources: laser
    obstacle_range: 2.5
    origin_z: 0.0
    raytrace_range: 3.0
    track_unknown_space: true
    unknown_threshold: 6
    z_resolution: 0.2
    z_voxels: 6
  obstacle_layer_footprint: {enabled: true}
  origin_x: 0.0
  origin_y: 0.0
  plugins:
  - {name: map_layer, type: 'costmap_2d::StaticLayer'}
  - {name: no_go_layer, type: 'costmap_2d::StaticLayer'}
  - {name: obstacle_layer, type: 'costmap_2d::VoxelLayer'}
  - {name: inflation_layer, type: 'costmap_2d::InflationLayer'}
  publish_frequency: 0.5
  resolution: 0.05
  robot_base_frame: /base_link
  robot_radius: 0.46
  rolling_window: false
  static_map: true
  transform_tolerance: 0.3
  update_frequency: 5.0
  width: 10
local_costmap:
  footprint: '[[-0.3,-0.32],[-0.52,-0.11],[-0.52,0.11],[-0.3,0.32],[-0.05,0.31],[0.05,0.25],[0.15,0.13],[0.2,0],[0.15,-0.13],[0.05,-0.25],[-0.05,-0.31]]'
  footprint_padding: 0.0001
  global_frame: /map
  height: 4
  inflation_layer: {cost_scaling_factor: 5.0, inflation_radius: 0.8}
  map_layer: {map_topic: /map}
  no_go_layer: {map_topic: /no_go_map}
  obstacle_layer:
    combination_method: 1
    enabled: true
    laser: {clearing: true, data_type: LaserScan, marking: true, sensor_frame: base_laser_link,
      topic: scan}
    mark_threshold: 0
    max_obstacle_height: 2.0
    observation_sources: laser
    obstacle_range: 2.5
    origin_z: 0.0
    raytrace_range: 3.0
    track_unknown_space: true
    unknown_threshold: 6
    z_resolution: 0.2
    z_voxels: 6
  obstacle_layer_footprint: {enabled: true}
  origin_x: 0.0
  origin_y: 0.0
  plugins:
  - {name: obstacle_layer, type: 'costmap_2d::VoxelLayer'}
  publish_frequency: 2.0
  resolution: 0.05
  robot_base_frame: /base_link
  robot_radius: 0.46
  rolling_window: true
  static_map: false
  transform_tolerance: 0.3
  update_frequency: 5.0
  width: 4
oscillation_distance: 0.5
oscillation_timeout: 0.0
planner_frequency: 0.5
planner_patience: 5.0
recovery_behavior_enabled: true
recovery_behaviors:
- {name: conservative_reset, type: clear_costmap_recovery/ClearCostmapRecovery}
- {name: aggressive_reset, type: clear_costmap_recovery/ClearCostmapRecovery}
restore_defaults: false
shutdown_costmaps: false
```

Yes, I am using policy execution from topo nav.

The easiest fix would be to set a timeout for the policy execution client but this feels somehow artificial.

**Comment by cdondrup** [comment]
Also, the goal is in free space. If the goal would be blocked, this would fail.

**Comment by bfalacerda** [comment]
I just tried calling move base directly via rviz and i get this:

```
[ WARN] [1446725936.535797813]: DWA planner failed to produce path.
[ INFO] [1446725936.633590127]: Got new plan
[ WARN] [1446725936.636505293]: DWA planner failed to produce path.
[ INFO] [1446725936.733484535]: Got new plan
[ WARN] [1446725936.735877074]: DWA planner failed to produce path.
[ WARN] [1446725936.833553834]: Clearing costmap to unstuck robot (2.000000m).
[ INFO] [1446725937.033437890]: Got new plan
[ WARN] [1446725937.035099355]: DWA planner failed to produce path.
[ WARN] [1446725937.133271232]: Clearing costmap to unstuck robot (0.000000m).
[ INFO] [1446725937.333269376]: Got new plan
[ WARN] [1446725937.336529415]: DWA planner failed to produce path.
[ERROR] [1446725937.433302700]: Aborting because a valid control could not be found. Even after executing all recovery behaviors
```

And then it stops. Can you try an rviz goal also? Do you see this error message in your case? If so, it aborted, and then was called again

**Comment by cdondrup** [comment]
OK, I tested a bit more. This seems to only happen in one of the test cases where the robot after clearing the costmap is bumping back a and forth by ~1cm all the time. Therefore it moves and DWA does not fail. In every other case, I also observe the behaviour you described. It's test map mb_test1 if you want to try.

**Comment by cdondrup** [comment]
Another problem, I start mon nav without a config file and get the following:

```
[WARN] [WallTime: 1446726452.096017] No config yaml file provided. MonitoredNavigation state machine will be initialized without recovery behaviours. To provide a config file do 'rosrun monitored_navigation monitored_nav.py path_to_config_file'. The strands config yaml file is located in monitored_navigation/config/strands.yaml
```

which I tought means that it would also not do the sleep and retry. Still it keeps on trying after it says aborted:

```
[ERROR] [1446726337.024717847]: Aborting because a valid control could not be found. Even after executing all recovery behaviors
[INFO] [WallTime: 1446726337.051787] State machine terminating 'NAVIGATION':'planner_failure':'not_recovered_without_help'
[INFO] [WallTime: 1446726337.052387] Concurrent state 'NAV_SM' returned outcome 'not_recovered_without_help' on termination.
[INFO] [WallTime: 1446726337.053695] Concurrent Outcomes: {'NAV_SM': 'not_recovered_without_help'}
[INFO] [WallTime: 1446726337.054235] State machine terminating 'MONITORED_NAV':'not_recovered_without_help':'not_recovered_without_help'
[INFO] [WallTime: 1446726337.054745] ABORTED
[INFO] [WallTime: 1446726337.303836] Intermediate -(move_base)-> End
[INFO] [WallTime: 1446726337.356830] State machine starting in initial state 'MONITORED_NAV' with userdata: 
    ['goal', 'result']
[INFO] [WallTime: 1446726337.357498] Concurrence starting with userdata: 
    ['goal']
[INFO] [WallTime: 1446726337.359362] State machine starting in initial state 'NAVIGATION' with userdata: 
    ['goal', 'n_fails']

```

So I get the above several times before my timeout of 1 minute kicks in and preempts it.

I am calling it like this:

``` python
self.client.send_goal(ExecutePolicyModeGoal(route=self._policy.route))
self.client.wait_for_result(timeout=rospy.Duration(self._timeout))
```

but this only returns after the timeout is reached and not when it fails for the first time. Is it supposed to retry navigation even if sleep and retry should not be there due to no config file?

**Comment by bfalacerda** [comment]
i see... maybe that has to do with the very aggressive clearing we have now: https://github.com/strands-project/strands_movebase/blob/indigo-devel/strands_movebase/strands_movebase_params/move_base_params.yaml#L15

This means we're clearing everything the robot cannot see from the costmap, which allows him to move a bit. maybe we should increase this value a bit?

**Comment by bfalacerda** [comment]
regarding the retrying, that's the policy executor which is configured to retry. You see that mon nav is outputting ABORTED there. there should be a param that you can change, @Jailander ?

**Comment by cdondrup** [comment]
Ah... redundancy... good...

I'll try the parameter. Thanks!

**Comment by Jailander** [comment]
hmm there should be one can't remember if not I'll make it ;)

**Comment by bfalacerda** [comment]
the sleep and retry is there to avoid asking for help too soon. the topo nav retry is there to make sure that after failing (and being pushed around during hellp for example) the robot checks its current waypoint and executes the appropriate action

**Comment by Jailander** [comment]
oh yeah I'll set this up with @cdondrup

**Comment by Jailander** [comment]
that parameter existed for topological navigation but not for execute_policy_mode a fix for this is in https://github.com/strands-project/strands_navigation/pull/275

**Comment by cburbridge** [comment]
@creuther The issue with the delay on the odometry channel or commanding the robot is not a problem with the ROS side, but the MIRA side. I have created a separate issue for it here: https://github.com/strands-project/scitos_drivers/issues/114

**Comment by cdondrup** [comment]
The navigation tests have been merged into strands_navigation. This has not been released yet, but have a look at the [README](https://github.com/strands-project/strands_navigation/blob/indigo-devel/topological_navigation/tests/README.md) and check out the current version of the repo to run tests in simulation and on the real robot.

**Comment by nilsbore** [comment]
Cool! Is it time for another hangout or should we just start running these tests IRL and then report back our findings?

**Comment by cdondrup** [comment]
You can try the tests first and let me know if you have any questions/suggestions. Still missing are the dynamic tests and also the exact definition of what kind of feedback we want from the test.

In general, I think a hangout would be good.

**Comment by marc-hanheide** [comment]
Agreed on the hangout. Can you initiate a doodle? Also, I have released strands_navigation again to include the tests.

**Comment by denisehe** [comment]
hello, 
well I followed your discussions - but actually all the technical details do not tell me that much.
is it possible that you film the tests so that I can see how the robot reacts to different test situations - so I can give you profound feedback if it is okay for the deployment or not :-)
Cheers from sunny and extremely hot vienna!

**Comment by cdondrup** [comment]
I sent an email to everyone who participated in the last hangout but in case someone new wants to join: http://doodle.com/poll/dchmfrzzwsca2qnk

Please fill in the doodle till the end of the week.

**Comment by cburbridge** [comment]
I have found what I suspect to be a bug in the scitos URDF: https://github.com/strands-project/scitos_common/pull/59. If this is an issue then it will have an effect on the movebase parameters....

**Comment by cdondrup** [comment]
Regarding the forking of ros navigation into our github mentioned in [here](https://github.com/strands-project/strands_movebase/issues/62#issuecomment-151664938), is the current navigation fork already @cburbridge 's version? Is this released already? I am currently changing the DWA to include my velocity costmaps so I would like to be able to contribute to our fork of the navigation framework. Can someone also pull the latest changes from upstream and enable issues on our fork please?

**Comment by marc-hanheide** [comment]
I enabled issues

**Comment by marc-hanheide** [comment]
I have also done the following on https://github.com/strands-project/navigation:
- merged the upstream of indigo-devel into our fork
- I have made indigo-devel the default branch
- I have enabled [devel](https://lcas.lincoln.ac.uk/jenkins/job/devel-indigo-navigation_strands_fork/) and [PR](https://lcas.lincoln.ac.uk/jenkins/job/pr-indigo-navigation_strands_fork/) build on jenkins for our fork (as for all our other repos)
- I have named our fork not "navigation", but "navigation_strands_fork" to avoid conflicts with the offical ones (This is only an internal name, we are not releasing Ubuntu packages from that fork yet, but once we do, we need to think of a name for it, or, better, a version prefix)

**Comment by cdondrup** [comment]
Thank you @marc-hanheide !

**Comment by marc-hanheide** [comment]
also changed all maintainers to `marc@hanheide.net` in all `package.xml` to make sure we on't harass the original maintainers when our builds fail. Feel free to change it to your name, when you do change things.

**Comment by marc-hanheide** [comment]
agreed in meeting today:
- @bfalacerda to make sure we only use strands_movebase parameters (https://github.com/strands-project/strands_morse/issues/144)
- put the correct params as discussed above in strands_movebase
- update the footprint
- make sure that all those PRs related to this are merged
- run the tests as described in https://github.com/strands-project/strands_navigation/blob/indigo-devel/topological_navigation/tests/README.md from a workspace that contains
  - https://github.com/strands-project/navigation
  - https://github.com/strands-project/strands_movebase
  - https://github.com/strands-project/strands_navigation

**Comment by nilsbore** [comment]
Sorry I wasn't there, had forgotten to put it in my calendar. Will follow up on this.

**Comment by nilsbore** [comment]
It seems that the moving of the laser and chest camera frames, together with the changing of the footprint, has improved navigation quite a bit in some aspects. What I've seen here is that the robot is traversing doors noticeably better than before.

Overall, when I ran navigation for a few hours the other day it seemed more robust but that is harder to pen down to something specific. Time to run the navigation tests =P.

**Comment by cdondrup** [comment]
Currently failing tests in simulation due to parameters:
- `test_static_1m_l_shaped_corridor`: After taking the bend the robot comes too close to one of the walls and shows the well known behaviour of just turning towards it and dying.
- `test_static_70cm_wide_door`: Still unsure if this will ever be solvable.
- `test_static_corridor_blocked_by_humans`: Robot plans through groups of humans (even though there is not enough space) and traps itself between them.
- `test_static_corridor_blocked_by_wheelchairs`: Robot plans through one of the wheelchairs because it only sees the wheels. This might be fixable using 3D obstacle avoidance. Nevertheless, there is not enough space for the robot to go between the wheels either.
- `test_static_facing_wall_10cm_distance_0_degrees_goal_behind`: Too close to wall for move base to get free.
- `test_static_facing_wall_10cm_distance_minus_45_degrees_goal_behind`: Too close to wall for move base to get free. Turns to +45 and then dies.
- `test_static_facing_wall_10cm_distance_plus_45_degrees_goal_behind`: Too close to wall for move base to get free. Turns to -45 and then dies.
- `test_static_human_on_end_point`: Not able to go back to start after failing to reach the end point.
- `test_static_wheelchair_on_end_point`: Not able to go back to start after failing to reach the end point.

**Comment by nilsbore** [comment]
Maybe it's time to switch `strands_morse` to `strands_movebase` and possibly add the chest camera to the simulation. The nice thing about the chest camera in simulation is that we can use the point cloud as-is, it won't be as expensive as the computations we do for the head cam to create the color point cloud and the images. What do you think?

**Comment by bfalacerda** [comment]
if navigation is being launched using strands_navigation.launch, then morse is also using strands_movebase. see https://github.com/strands-project/strands_morse/issues/144

**Comment by bfalacerda** [comment]
regarding chest cam, we could add it yes, i guess it'll be important for the wheelchairs

**Comment by Jailander** [comment]
yes that would be brilliant, however I think jenkins can only run simulations in the fast mode without cameras (right @marc-hanheide ?) but we could still run the tests independently using a flag like for the real robot

**Comment by nilsbore** [comment]
Ok, I was just looking at this launch file: https://github.com/strands-project/strands_morse/blob/indigo-devel/mba/launch/move_base_arena_nav.launch .

**Comment by nilsbore** [comment]
I'll have a look at adding the chest cam sometime during this week.

**Comment by marc-hanheide** [comment]
Can we please leave the default without chestcam in simulation. Many students and other people  (including jenkins) don't have a suitable GPU and can't run the full Morse. Of course it would be great to add chestcam and a suitable flag.

**Comment by nilsbore** [comment]
Of course, it will be an option.

**Comment by cdondrup** [comment]
Once we have a navigation set-up that runs in simulation and on the robot, I can add that to the tests and have a flag as well, yes.

**Comment by cdondrup** [comment]
We did the first real robot tests based on the test scenarios in simulation. You find a few videos here: https://lcas.lincoln.ac.uk/owncloud/index.php/s/GppNjlMSMyaHTVf#

Short summary or what isn't working:
- the robot starting 10cm away from the wall. Doesn't even move regardless of the angle it is facing it at.
- the middle waypoint blocked by an obsctacle, robot doesn't even try to move. According to @Jailander that should have been fixed and the robot should try to go as close as possible to that point at least but we didn't observe that.
- the last waypoint is blocked. The robot doesn't even try to move towards the last waypoint after reaching the intermediate one. Not sure if this should be changed or is actually a good behaviour.

**Comment by marc-hanheide** [comment]
Many thanks, @Jailander and @cdondrup for getting going on this.
I think it's now time we reconvene on hangout to replicate this at the various sites and to discuss the improvements possible to make it pass all those tests. The latter seem to be mostly related to topological navigation, so @Jailander and @bfalacerda are probably the ones to look at those. 

Question to @cdondrup: I presume this was based on the latest set of parameters etc. Just for the record, can you comment and link the exact commits/released versions you have been using for your tests. Also, given you previous experience, would you say this was already better than what we had in the past, as we had some improvements (accelerations, laser position, etc) implemented already?

In any case, I put a doodle for the next meeting out: https://doodle.com/poll/mtynvcxw23fnsmdp

Please pick you slots quickly, so we can have the meeting in short term.

**Comment by bfalacerda** [comment]
i'd expect the robot to at least try to get closer to the wp in the second situation, even if he doesnt manage to get through it... I'll also try that and see what happens. Is `float64 xy_goal_tolerance` of the waypoints set to something? 

Regarding the last one, that's the expected behaviour and I also think it's acceptable. we cant really hope to do anything better than that.

I have a suggestion regarding this intermediate waypoints issue, and also all the weird trajectories the robot takes when following a policy between nodes: the current version changes goal once we get into the influence area of a node, and the next action is also move_base/human_aware. I suggest a two step lookahead instead, i.e., if we have a waypoint sequence A->B->C->D, all with move_base, once we get to the influence area of B, we send a move_base goal to D. This will surely make the robot's movement much smoother (e.g., it'll probably avoid the weird nav in the AAF corridors), and will also make it more robust to occupied intermediate nodes. The only drawback I see is when the robot is doing B->C->D, and ends up not getting into C's influence area, for example because of an obstacle. We'd have to be careful with that during execution, and with the nav stats.

@Jailander does this sound sensible?

Another option would be to change the move_base goal immediately after the current target becomes the closest_node. That would have the same advantages and issues more or less.

**Comment by Jailander** [comment]
@bfalacerda that could be possible I'm unsure of the consequences of this in terms of making the robot follow the actual topological route and not taking another route, I'll put it in as soon as I have time on a separate branch and test it in simulation.

About the XY tolerance you are right this this should have worked, I'll rerun the tests today

**EDIT**: we already change the move_base goal immediately after the current target becomes the closest_node (forgot to write this)

**Comment by cburbridge** [comment]
This needs fixing in the simulation as it might have an effect on the testing results https://github.com/strands-project/strands_morse/issues/143

**Comment by bfalacerda** [comment]
@Jailander when was that changed? is nav smoother now?

**Comment by Jailander** [comment]
oops no my mistake, just checked, I did that in the ICRA branch but never pushed it upstream as that branch is substantially different from our main system, the behaviour it had was worse, the reason for that is that when the robot changed goal sooner it was going faster so it stopped harder (but in that case there was a bigger time gap between move base goals because we switched maps too), I am not sure if we really should try it in the main system too, do you think its necessary?

**Comment by bfalacerda** [comment]
I think he shouldn't stop at all, if you send a new goal to move_base it usually changes quite smoothly. We have deceleration now because he was getting close to the goal before receiving the new one, but if we send new goals from further away it should be smoother. Why does he stop when he gets a new goal?

**Comment by Jailander** [comment]
hmm in that case it could have been because there was no (metric) map for some time, I'll test it on normal conditions and tell you.

**Comment by marc-hanheide** [comment]
Minutes of Meeting:
- We only use the parameters in strands_movebase
- We re-released strands_movebase, strands_navigation, scitos_common to be up to date
- Then everybody should replicate the tests according to the video¯

**Comment by marc-hanheide** [comment]
All releases are out. Time to update and run the tests.

**Comment by nilsbore** [comment]
Unfortunately, it looks like Rosie's PCB power board has fried again, meaning we have no main computer atm. We'll see what happens but it likely won't be possible to run tests this week.

**Comment by marc-hanheide** [comment]
Ah, bugger... OK, we'll have to see what we can do here and hopefully Birmingham @kunzel @bfalacerda will be able to run on their site?

**Comment by bfalacerda** [comment]
yes, we have student that's going to help, I'm meeting him today to set up the testing. Just to confirm, the idea now is to run the static tests one by one and report the result right?

**Comment by bfalacerda** [comment]
So we have the robot ready to run tests, we just need to make some arenas resembling the scenarios.

~~@Jailander @cdondrup the output from the "push robot to start location" always tells me I'm 2 meters away from the goal.~~ Ignore this, the topo node pose wasn't being updated properly for some reason
